File: app_flutter_horario_mobile/lib/domain/curso.dart
import 'package:app_flutter_horario_mobile/domain/coordenador.dart';
import 'package:app_flutter_horario_mobile/domain/modulo.dart';

class Curso {
  int? _id;
  String? _nome;
  String? _nivelEnsino;
  String? _qdtPeriodos;
  String? _modalidade;
  int? _cargaTotal;
  Coordenador? _coordenador;
  List<Modulo>? _modulos;

  Curso(
      {int? id,
      String? nome,
      String? nivelEnsino,
      String? qdtPeriodos,
      String? modalidade,
      int? cargaTotal,
      Coordenador? coordenador,
      List<Modulo>? modulos});

  int? get id => this._id;

  set id(int? value) => this._id = value;

  get nome => this._nome;

  set nome(value) => this._nome = value;

  get nivelEnsino => this._nivelEnsino;

  set nivelEnsino(value) => this._nivelEnsino = value;

  get qdtPeriodos => this._qdtPeriodos;

  set qdtPeriodos(value) => this._qdtPeriodos = value;

  get modalidade => this._modalidade;

  set modalidade(value) => this._modalidade = value;

  get cargaTotal => this._cargaTotal;

  set cargaTotal(value) => this._cargaTotal = value;

  get coordenador => this._coordenador;

  set coordenador(value) => this._coordenador = value;

  get modulos => this._modulos;

  set modulos(value) => this._modulos = value;
}

File: backend/systempus/src/main/java/br/com/systempus/systempus/domain/Curso.java
package br.com.systempus.systempus.domain;

import java.util.List;

import com.fasterxml.jackson.annotation.JsonBackReference;
import com.fasterxml.jackson.annotation.JsonIdentityInfo;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonManagedReference;
import com.fasterxml.jackson.annotation.ObjectIdGenerators;

import br.com.systempus.systempus.domain.enumerador.Modalidade;
import br.com.systempus.systempus.domain.enumerador.NivelEnsino;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToMany;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.Table;
import jakarta.validation.constraints.NotNull;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Entity
@Table(name = "curso")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@JsonIdentityInfo(generator = ObjectIdGenerators.PropertyGenerator.class, property = "id")
public class Curso {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer id;

    @NotNull(message = "Campo obrigatório")
    private String nome;

    @NotNull(message = "Campo obrigatório")
    @Enumerated(EnumType.ORDINAL)
    private NivelEnsino nivelEnsino;

    @NotNull(message = "Campo obrigatório")
    private Integer qtdPeriodos;

    @NotNull(message = "Campo obrigatório")
    @Enumerated(EnumType.ORDINAL)
    private Modalidade modalidade;

    private Integer cargaTotal;

    // @JsonManagedReference(value = "curso_modulos")
    @OneToMany(mappedBy = "curso")
    private List<Modulo> modulos;

    @JsonBackReference(value = "curso_coordenador")
    @ManyToOne
    @JoinColumn(name = "coordenador_id", nullable = true)
    private Coordenador coordenador;

    @ManyToMany (mappedBy = "cursos")
    private List<Professor> professores;

    @OneToMany(mappedBy = "curso")
    private List<Periodo> periodos;
    
    @Override
    public String toString() {
        return "Curso [id=" + id + ", nome=" + nome + ", nivelEnsino=" + nivelEnsino + ", qtdPeriodos=" + qtdPeriodos
                + ", modalidade=" + modalidade + ", cargaTotal=" + cargaTotal + ", modulo=" + modulos
                + ", coordenador=" + coordenador + "]";
    }


    
}
